include/ginkgo/core/base/mpi.hpp

```diff
@@ -0,0 +1,1221 @@
+    {
+        int flag = 0;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Finalized(&flag));
+        return flag;
+    }
+
+    static bool is_initialized()
+    {
+        int flag = 0;
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Initialized(&flag));
+        return flag;
+    }
+
+    /**
+     * Return the provided thread support.
+     *
+     * @return the provided thread support
+     */
+    int get_provided_thread_support() { return provided_thread_support_; }
+
+    /**
+     * Call MPI_Init_thread and initialize the MPI environment
+     *
+     * @param argc  the number of arguments to the main function.
+     * @param argv  the arguments provided to the main function.
+     * @param thread_t  the type of threading for initialization. See
+     *                  @thread_type
+     */
+    environment(int& argc, char**& argv,
+                const thread_type thread_t = thread_type::serialized)
+    {
+        this->required_thread_support_ = static_cast<int>(thread_t);
+        GKO_ASSERT_NO_MPI_ERRORS(
+            MPI_Init_thread(&argc, &argv, this->required_thread_support_,
+                            &(this->provided_thread_support_)));
+    }
+
+    /**
+     * Call MPI_Finalize at the end of the scope of this class.
+     */
+    ~environment() { MPI_Finalize(); }
+
+    environment(const environment&) = delete;
+    environment(environment&&) = delete;
+    environment& operator=(const environment&) = delete;
+    environment& operator=(environment&&) = delete;
+
+private:
+    int required_thread_support_;
+    int provided_thread_support_;
+};
+
+
+/**
+ * Returns if GPU aware functionality has been enabled
+ */
+static bool is_gpu_aware()
+{
+#if GKO_HAVE_GPU_AWARE_MPI
+    return true;
+#else
+    return false;
+#endif
+}
+
+
+namespace {
+
+
+/**
+ * A deleter class that calls MPI_Comm_free on the owning MPI_Comm object and
+ * deletes the underlying comm ptr
+ */
+class comm_deleter {
+public:
+    using pointer = MPI_Comm*;
+    void operator()(pointer comm) const
+    {
+        GKO_ASSERT(*comm != MPI_COMM_NULL);
+        GKO_ASSERT_NO_MPI_ERRORS(MPI_Comm_free(comm));
+        delete comm;
+    }
+};
+
+
+}  // namespace
+
+
+/**
+ * A communicator class that takes in the given communicator and duplicates it
+ * for our purposes. As the class or object goes out of scope, the communicator
+ * is freed.
+ */
+class communicator {
+public:
+    /**
+     * Non-owning constructor for an existing communicator of type MPI_Comm. The
+     * MPI_Comm object will not be deleted after the communicator object has
+     * been freed and an explicit MPI_Comm_free needs to be called on the
+     * original MPI_Comm object.
+     *
+     * @param comm The input MPI_Comm object.
+     */
+    communicator(const MPI_Comm& comm)
+    {
+        this->comm_.reset(new MPI_Comm(comm));
+        this->size_ = get_num_ranks();
+        this->rank_ = get_my_rank();
+        this->node_local_rank_ = get_node_local_rank();
```

Comment:
Returning these values is a pretty trivial operation inside OpenMPI and MPICH, so do we really need to cache them? For node-local rank, I guess this is something to query when setting up a generic executor, so maybe ignore the overhead and fetch the information each time, since this is not a common operation at all?

Reply:
Is there any disadvantage in caching them ?

Reply:
more complex in constructor calls, larger object size

Reply:
It's just an integer, and we'll probably have only one communicator in most program runs. On the other hand, splitting the communicator and getting the local rank might need queries to hwloc etc. internally. IMO it's a good idea to cache this. If anything, the size and rank need not be cached as they're probably already cached in the MPI_Comm anyway (just saying; it's fine to keep them).

Reply:
I was talking mostly about size and rank. Do you expect to need to compute the local rank regularly? I can imagine two scenarios where this is important, 1. when running multi-GPU setups with all GPUs visible to all ranks, and 2. when doing hierarchical intra- and inter-node communication. The former case needs this value only once when setting up the GPUs, and the latter case would require a node-local communicator separately, from which we could then get the local rank directly. It may be a bit surprising that the construction of a communicator (or even just wrapping one) involves a collective call.